File: leetcode_python/322. Coin Change.py
class Solution:
    def coinChange(self, coins: [int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0
        for money in range(1, amount + 1):
            for coin in coins:
                if coin <= money:
                    dp[money] = min(dp[money], dp[money - coin])
            dp[money] += 1
        return dp[amount] if dp[amount] <= amount else -1

File: leetcode_java/322. Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        dp[0] = 0;
        for(int money = 1; money <= amount; money++) {
            for(int coin : coins) {
                if(coin <= money)
                    dp[money] = Math.min(dp[money], dp[money - coin]);
            }
            dp[money]++;
        }
        return dp[amount] <= amount ? dp[amount] : -1;
    }
}